File: native/src/wn_stub.c
/*
 * web-node — native → WASM 迁移的冒烟桩（M115）
 *
 * 这不是任何真实 Node binding 的实现，只用来验证 M115 的接入缝：
 *   1. 真 C 源码（wasi-sdk clang，`--target=wasm32-wasip1`）能编出 wasm32-wasi 模块；
 *   2. 模块导出函数 + 线性内存，宿主侧能 `WebAssembly.instantiate` 后调用；
 *   3. `malloc`/`free`（wasi-libc）可用，宿主能透过线性内存读回数据。
 *
 * 导出用 `__attribute__((export_name(...)))` 精确声明——这样不必给链接器传
 * `--export-all`，libc 内部符号（malloc/free/…）不会泄进导出表。
 *
 * M116 起，真实 binding（zlib、histogram、crypto…）会把 `deps/` 里的上游
 * C/C++ 源码编成同类模块，并经同一个加载器接入 `internalBinding()`。
 */
#include <stdint.h>
#include <stdlib.h>

/* 最简调用链：验证「导出函数 → 宿主调用」通。 */
__attribute__((export_name("wn_stub_add"), used))
int32_t wn_stub_add(int32_t a, int32_t b) {
  return a + b;
}

/*
 * 分配 `len` 字节并写入 `seed, seed+1, …`（uint8 回绕），返回指针。
 * 宿主读回线性内存即可验证 malloc 与内存 ABI。
 */
__attribute__((export_name("wn_stub_alloc_fill"), used))
uint8_t *wn_stub_alloc_fill(int32_t len, uint8_t seed) {
  uint8_t *p = (uint8_t *)malloc((size_t)len);
  if (p == NULL) {
    return NULL;
  }
  for (int32_t i = 0; i < len; i++) {
    p[i] = (uint8_t)(seed + i);
  }
  return p;
}

/* 释放 `wn_stub_alloc_fill` 返回的指针。 */
__attribute__((export_name("wn_stub_free"), used))
void wn_stub_free(void *p) {
  free(p);
}

/*
 * 返回指向只读数据段的静态字符串。宿主用 `TextDecoder` 读回，
 * 验证「数据段 + 指针 → 字符串」这条往返。
 */
__attribute__((export_name("wn_stub_version"), used))
const char *wn_stub_version(void) {
  return "web-node wasm stub 1";
}

File: src/node-runtime/bindings/wn_stub.ts
/**
 * `wn_stub` —— native → WASM 接入缝的冒烟 binding（M115）。
 *
 * 它**不实现任何真实 Node binding**，只证明这条链通了：
 *   真 C 源码 → wasi-sdk（wasm32-wasip1）→ 实例化 → `internalBinding('wn_stub')` 取用。
 * 覆盖了导出函数、线性内存、`malloc`/`free` 三件事。M116 起，zlib / histogram /
 * crypto 会以同样方式接进来，只是导出表和包装更厚。
 *
 * 取用方式与真 Node 一致：绑定工厂在 Realm 构造期同步运行，从
 * {@link ../wasm/registry} 里同步取已实例化的模块（worker 启动时已 fetch + 编译好）。
 */
import type { BindingContext } from './context';
import { wasmModule } from '../wasm/registry';

/** 该模块的导出表（未加载则抛，绝不静默返回 undefined）。 */
function exports(): Record<string, unknown> {
  return wasmModule('wn_stub');
}

/** 模块的线性内存。 */
function memory(): WebAssembly.Memory {
  return exports().memory as WebAssembly.Memory;
}

export function wnStubBinding(_ctx: BindingContext): Record<string, unknown> {
  return {
    /** 两数相加：最简「导出函数 → 宿主调用」链。 */
    add(a: number, b: number): number {
      const fn = exports().wn_stub_add as (a: number, b: number) => number;
      return fn(a | 0, b | 0);
    },

    /** 读只读数据段里的静态字符串（数据段 + 指针 → 字符串）。 */
    version(): string {
      const ptr = (exports().wn_stub_version as () => number)();
      const bytes = new Uint8Array(memory().buffer);
      let end = ptr;
      while (bytes[end] !== 0) end++;
      return new TextDecoder().decode(bytes.subarray(ptr, end));
    },

    /**
     * `malloc` → 写入 → 拷回 → `free`：验证内存 ABI。
     * 返回的是**拷贝**（不是视图），所以 `free` 之后依然有效。
     */
    roundTrip(len: number, seed: number): Uint8Array {
      const ex = exports();
      const ptr = (ex.wn_stub_alloc_fill as (n: number, s: number) => number)(len | 0, seed & 0xff);
      if (ptr === 0) throw new Error('wn_stub: allocation failed');
      try {
        return new Uint8Array(memory().buffer, ptr, len).slice();
      } finally {
        (ex.wn_stub_free as (p: number) => void)(ptr);
      }
    },
  };
}
